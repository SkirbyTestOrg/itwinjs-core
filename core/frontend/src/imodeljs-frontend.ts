/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/

export * from "./tools/AccuDrawTool";
export * from "./tools/EditManipulator";
export * from "./tools/ElementSetTool";
export * from "./tools/EventController";
export * from "./tools/MeasureTool";
export * from "./tools/SelectTool";
export * from "./tools/ToolAdmin";
export * from "./tools/Tool";
export * from "./tools/ViewTool";
export * from "./tools/ClipViewTool";
export * from "./tools/PrimitiveTool";
export * from "./tools/IdleTool";
export * from "./tools/ToolAssistance";
export * from "./AccuDraw";
export * from "./AccuSnap";
export * from "./AuxCoordSys";
export * from "./DevTools";
export * from "./FrontendRequestContext";
export * from "./CategorySelectorState";
export * from "./ContextRealityModelState";
export * from "./SpatialClassification";
export * from "./DisplayStyleState";
export * from "./ElementLocateManager";
export * from "./EmphasizeElements";
export * from "./EntityState";
export * from "./FuzzySearch";
export * from "./GeoServices";
export * from "./HitDetail";
export * from "./IModelConnection";
export * from "./ImageUtil";
export * from "./FrontendLoggerCategory";
export * from "./Marker";
export * from "./ModelSelectorState";
export * from "./ModelState";
export * from "./NotificationManager";
export * from "./plugin/Plugin";
export * from "./plugin/PluginUiManager";
export * from "./RelativePosition";
export * from "./RenderScheduleState";
export * from "./SelectionSet";
export * from "./Sheet";
export * from "./Sprites";
export * from "./StandardView";
export * from "./SubCategoriesCache";
export * from "./TentativePoint";
export * from "./QuantityFormatter";
export * from "./ViewContext";
export * from "./ViewManager";
export * from "./Viewport";
export * from "./ViewState";
export * from "./IModelApp";
export * from "./NoRenderApp";
export * from "./tile/TileAdmin";
export * from "./tile/Tile";
export * from "./tile/TileTree";
export * from "./tile/WebMapTileTree";
export * from "./tile/BackgroundTerrainTileTree";
export * from "./tile/CesiumWorldTerrainTileTree";
export * from "./tile/BingElevation";
export * from "./render/FeatureSymbology";
export * from "./render/GraphicBuilder";
export * from "./render/MockRender";
export * from "./render/System";
export * from "./render/webgl/Target";
export * from "./RenderCompatibility";
export * from "./oidc/OidcBrowserClient";
export * from "./TerrainProvider";
export * from "./properties/Description";
export * from "./properties/EditorParams";
export * from "./properties/Record";
export * from "./properties/Value";
export * from "./properties/ToolSettingsValue";
export * from "./properties/PrimitiveTypes";
export * from "./properties/BaseQuantityDescription";
export * from "./properties/AngleDescription";
export * from "./properties/LengthDescription";
export * from "./SolarCalculate";

// Set the version number so it can be found at runtime. BUILD_SEMVER is replaced at build time by the webpack DefinePlugin.
declare var BUILD_SEMVER: string;
declare var BUILD_TYPE: string;
if ((typeof (BUILD_SEMVER) !== "undefined") && (typeof window !== "undefined") && window) {
  if (!(window as any).iModelJsVersions)
    (window as any).iModelJsVersions = new Map<string, string>();
  (window as any).iModelJsVersions.set("imodeljs-frontend", BUILD_SEMVER);
  (window as any).iModelJsVersions.set("buildType", BUILD_TYPE);
}

/** @docs-package-description
 * The ($frontend) package always runs in a web browser. It contains classes for [querying iModels and showing views]($docs/learning/frontend/index.md).
 */

/**
 * @docs-group-description IModelApp
 * Classes for configuring and administering an iModel.js application.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
/**
 * @docs-group-description IModelConnection
 * Classes for working with a connection to an [iModel briefcase]($docs/learning/IModels.md)
 */
/**
 * @docs-group-description ElementState
 * Classes for working with the *state* of Elements in the frontend.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
/**
 * @docs-group-description ModelState
 * Classes for working with the *state* of Models in the frontend.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
/**
 * @docs-group-description Tools
 * Classes for [working with Tools]($docs/learning/frontend/Tools.md)
 */
/**
 * @docs-group-description Views
 * Classes for [working with Views]($docs/learning/frontend/Views.md)
 */
/**
 * @docs-group-description LocatingElements
 * Classes for locating and snapping to elements in views.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
/**
 * @docs-group-description AccuDraw
 * AccuDraw provides helpful assistance for creating and modifying elements in a view.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
/**
 * @docs-group-description Notifications
 * Notifications provide feedback to the user of something of interest.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
/**
 * @docs-group-description Plugins
 * Classes for creating and managing runtime [Plugins]($docs/learning/frontend/Plugins.md)
 */
/**
 * @docs-group-description Properties
 * Classes for working with property records and descriptions.
 */
/**
 * @docs-group-description Rendering
 * Classes for rendering the contents of views.
 */
/**
 * @docs-group-description SelectionSet
 * Classes for working with the set of selected elements.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
