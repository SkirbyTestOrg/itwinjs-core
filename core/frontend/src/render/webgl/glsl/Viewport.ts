/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module WebGL */

import { ShaderBuilder, VariableType, VertexShaderBuilder } from "../ShaderBuilder";
import { Matrix4 } from "../Matrix";
import { ViewRect } from "../../../Viewport";
import { addModelViewProjectionMatrix } from "./Vertex";
import { addRenderPass } from "./RenderPass";

export function addViewport(shader: ShaderBuilder) {
  shader.addUniform("u_viewport", VariableType.Vec4, (prog) => {
    prog.addProgramUniform("u_viewport", (uniform, params) => {
      const rect = params.target.viewRect;
      const vp: number[] = [rect.left, rect.bottom, rect.width, rect.height];
      uniform.setUniform4fv(vp);
    });
  });
}

function computeViewportTransformation(viewRect: ViewRect, nearDepthRange: number, farDepthRange: number): Matrix4 {
  const x = viewRect.left;
  const y = viewRect.top;
  const width = viewRect.width;
  const height = viewRect.height;

  const halfWidth = width * 0.5;
  const halfHeight = height * 0.5;
  const halfDepth = (farDepthRange - nearDepthRange) * 0.5;

  const column0Row0 = halfWidth;
  const column1Row1 = halfHeight;
  const column2Row2 = halfDepth;
  const column3Row0 = x + halfWidth;
  const column3Row1 = y + halfHeight;
  const column3Row2 = nearDepthRange + halfDepth;
  const column3Row3 = 1.0;

  const mat = Matrix4.fromValues(
    column0Row0, 0.0, 0.0, column3Row0,
    0.0, column1Row1, 0.0, column3Row1,
    0.0, 0.0, column2Row2, column3Row2,
    0.0, 0.0, 0.0, column3Row3);

  return mat;
}

export function addViewportTransformation(shader: ShaderBuilder) {
  shader.addUniform("u_viewportTransformation", VariableType.Mat4, (prog) => {
    prog.addGraphicUniform("u_viewportTransformation", (uniform, params) => {
      uniform.setMatrix4(computeViewportTransformation(params.target.viewRect, 0.0, 1.0));
    });
  });
}

const modelToWindowCoordinates = `
vec4 modelToWindowCoordinates(vec4 position, vec4 next) {
  if (kRenderPass_ViewOverlay == u_renderPass || kRenderPass_Background == u_renderPass) {
    vec4 q = u_mvp * position;
    q.xyz /= q.w;
    q.xyz = (u_viewportTransformation * vec4(q.xyz, 1.0)).xyz;
    return q;
  }

  // Negative values are in front of the camera (visible).
  float s_maxZ = -u_frustum.x;            // use -near (front) plane for segment drop test since u_frustum's near & far are pos.
  vec4  q = u_mv * position;              // eye coordinates.
  vec4  n = u_mv * next;

  if (q.z > s_maxZ) {
    if (n.z > s_maxZ)
      return vec4(0.0, 0.0,  1.0, 0.0);   // Entire segment behind eye.

    float t = (s_maxZ - q.z) / (n.z - q.z);

    q.x += t * (n.x - q.x);
    q.y += t * (n.y - q.y);
    q.z = s_maxZ;                       // q.z + (s_maxZ - q.z) * (s_maxZ - q.z) / n.z - q.z
  }
  q = u_proj * q;
  q.xyz /= q.w;                           // normalized device coords
  q.xyz = (u_viewportTransformation * vec4(q.xyz, 1.0)).xyz; // window coords
  return q;
  }
`;

export function addModelToWindowCoordinates(vert: VertexShaderBuilder) {
  addModelViewProjectionMatrix(vert);
  addViewportTransformation(vert);
  addRenderPass(vert);
  vert.addFunction(modelToWindowCoordinates);
}
