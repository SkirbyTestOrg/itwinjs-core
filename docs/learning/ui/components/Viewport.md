# Viewport

## API Reference

* [Viewport]($ui-components:Viewport)
