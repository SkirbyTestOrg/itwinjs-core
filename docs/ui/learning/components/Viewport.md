# Viewport

## API Reference

* [Viewport]($components:Viewport)
