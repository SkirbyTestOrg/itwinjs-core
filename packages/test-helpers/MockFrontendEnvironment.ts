/*---------------------------------------------------------------------------------------------
|  $Copyright: (c) 2018 Bentley Systems, Incorporated. All rights reserved. $
 *--------------------------------------------------------------------------------------------*/
(global as any).WebGLRenderingContext = require("gl"); // tslint:disable-line:no-var-requires
(global as any).XMLHttpRequest = require("xmlhttprequest").XMLHttpRequest; // tslint:disable-line:no-var-requires
