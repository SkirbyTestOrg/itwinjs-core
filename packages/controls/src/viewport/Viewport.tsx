/*---------------------------------------------------------------------------------------------
|  $Copyright: (c) 2018 Bentley Systems, Incorporated. All rights reserved. $
 *--------------------------------------------------------------------------------------------*/
import * as React from "react";
import { Id64Props } from "@bentley/bentleyjs-core";
import { IModelConnection } from "@bentley/imodeljs-frontend";
import { ViewportComponent } from "@bentley/ui-components";
import SelectionHandler from "./SelectionHandler";

/**
 * Props for [[Viewport]] component.
 */
export interface Props {
  /** iModel to display */
  imodel: IModelConnection;

  /** ID of default view definition which should be displayed */
  viewDefinitionId: Id64Props;

  /**
   * ID of a presentation ruleset used to determine what should
   * get highlighted in the viewport when unified selection changes
   */
  rulesetId: string;

  /** @hidden */
  selectionHandler?: SelectionHandler;
}

/**
 * Viewport component which participates in unified selection.
 */
export default class Viewport extends React.Component<Props> {

  private _selectionHandler?: SelectionHandler;

  private static getSelectionHandlerFromProps(props: Props): SelectionHandler {
    const handler = props.selectionHandler
      ? props.selectionHandler : new SelectionHandler(props.imodel, props.rulesetId);
    return handler;
  }

  public componentDidMount() {
    this._selectionHandler = Viewport.getSelectionHandlerFromProps(this.props);
  }

  public componentWillUnmount() {
    if (this._selectionHandler) {
      this._selectionHandler.dispose();
      this._selectionHandler = undefined;
    }
  }

  public componentDidUpdate() {
    if (this._selectionHandler) {
      this._selectionHandler.imodel = this.props.imodel;
      this._selectionHandler.rulesetId = this.props.rulesetId;
    }
  }

  public get selectionHandler(): SelectionHandler {
    if (!this._selectionHandler)
      throw new Error("Component is not mounted");
    return this._selectionHandler;
  }

  public render() {
    return (<ViewportComponent
      imodel={this.props.imodel}
      viewDefinitionId={this.props.viewDefinitionId}
    />);
  }
}
