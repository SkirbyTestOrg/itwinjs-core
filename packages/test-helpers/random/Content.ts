/*---------------------------------------------------------------------------------------------
|  $Copyright: (c) 2017 Bentley Systems, Incorporated. All rights reserved. $
 *--------------------------------------------------------------------------------------------*/
import * as faker from "faker";
import * as c from "../../common/src/content";
import { createRandomECClassInfo, createRandomRelationshipPath } from "./EC";
import { nullable } from "./Misc";

const createRandomSelectClassInfo = (): c.SelectClassInfo => {
  return {
    selectClassInfo: createRandomECClassInfo(),
    isSelectPolymorphic: faker.random.boolean(),
    pathToPrimaryClass: createRandomRelationshipPath(),
    relatedPropertyPaths: [createRandomRelationshipPath(1), createRandomRelationshipPath(1)],
  } as c.SelectClassInfo;
};

export const createRandomCategory = (): c.CategoryDescription => {
  return {
    name: faker.random.word(),
    label: faker.random.words(),
    description: faker.lorem.sentence(),
    priority: faker.random.number(),
    expand: faker.random.boolean(),
  } as c.CategoryDescription;
};

export const createRandomPrimitiveTypeDescription = (): c.TypeDescription => {
  return {
    valueFormat: c.PropertyValueFormat.Primitive,
    typeName: faker.database.type(),
  } as c.PrimitiveTypeDescription;
};

export const createRandomEditorDescription = (): c.EditorDescription => {
  return {
    name: faker.random.word(),
  } as c.EditorDescription;
};

const createRandomPrimitiveFieldJson = () => {
  return {
    category: createRandomCategory(),
    name: faker.random.word(),
    label: faker.random.words(),
    type: createRandomPrimitiveTypeDescription(),
    isReadonly: faker.random.boolean(),
    priority: faker.random.number(),
    editor: nullable(createRandomEditorDescription),
  };
};

export const createRandomPrimitiveField = (): c.Field => {
  return c.Field.fromJSON(createRandomPrimitiveFieldJson())!;
};

export const createRandomDescriptorJson = (displayType?: string) => {
  const selectClasses = [createRandomSelectClassInfo(), createRandomSelectClassInfo()];
  const fields = [createRandomPrimitiveFieldJson(), createRandomPrimitiveFieldJson(), createRandomPrimitiveFieldJson()];
  return {
    connectionId: faker.random.uuid(),
    inputKeysHash: faker.random.uuid(),
    contentOptions: faker.random.objectElement(),
    displayType: displayType || faker.lorem.words(),
    selectClasses,
    fields,
    contentFlags: 0,
  };
};

export const createRandomDescriptor = (displayType?: string): c.Descriptor => {
  return c.Descriptor.fromJSON(createRandomDescriptorJson(displayType))!;
};
