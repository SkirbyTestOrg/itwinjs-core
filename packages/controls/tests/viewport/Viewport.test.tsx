/*---------------------------------------------------------------------------------------------
|  $Copyright: (c) 2018 Bentley Systems, Incorporated. All rights reserved. $
 *--------------------------------------------------------------------------------------------*/
import "@helpers/MockFrontendEnvironment";
import * as React from "react";
import { expect } from "chai";
import { mount, shallow } from "enzyme";
import * as faker from "faker";
import * as moq from "@helpers/Mocks";
import { IModelConnection, SelectionSet, ViewState3d, IModelApp } from "@bentley/imodeljs-frontend";
import { ECPresentation, SelectionManager, SelectionChangeEvent } from "@bentley/ecpresentation-frontend/lib";
import ViewportSelectionHandler from "@src/viewport/SelectionHandler";
import Viewport from "@src/viewport/Viewport";
import { createRandomId } from "@helpers/random/Misc";
import "@helpers/Snapshots";

describe("Viewport", () => {

  before(() => {
    IModelApp.startup();
  });
  after(() => {
    IModelApp.shutdown();
  });

  const viewDefinitionId = createRandomId();
  const imodelMock = moq.Mock.ofType<IModelConnection>();
  const selectionHandlerMock = moq.Mock.ofType<ViewportSelectionHandler>();
  beforeEach(() => {
    const viewsMock = moq.Mock.ofInstance<IModelConnection.Views>(new IModelConnection.Views(imodelMock.object));
    viewsMock.setup((views) => views.load(moq.It.isAny())).returns(async () => moq.Mock.ofType<ViewState3d>().object);
    selectionHandlerMock.reset();
    imodelMock.reset();
    imodelMock.setup((imodel) => imodel.selectionSet).returns((imodel) => new SelectionSet(imodel));
    imodelMock.setup((imodel) => imodel.views).returns(() => viewsMock.object);
  });

  it("mounts", () => {
    mount(<Viewport
      imodel={imodelMock.object}
      rulesetId={faker.random.word()}
      viewDefinitionId={viewDefinitionId}
      selectionHandler={ selectionHandlerMock.object }
    />);
  });

  it("renders correctly", () => {
    expect(shallow(<Viewport
      imodel={imodelMock.object}
      rulesetId={faker.random.word()}
      viewDefinitionId={viewDefinitionId}
      selectionHandler={selectionHandlerMock.object}
    />)).to.matchSnapshot();
  });

  describe("selectionHandler", () => {

    it("creates default implementation when not provided through props", () => {
      const selectionManagerMock = moq.Mock.ofType<SelectionManager>();
      selectionManagerMock.setup((x) => x.selectionChange).returns(() => new SelectionChangeEvent());
      ECPresentation.selection = selectionManagerMock.object;

      const rulesetId = faker.random.word();

      const viewport = shallow(<Viewport
        imodel={imodelMock.object}
        rulesetId={rulesetId}
        viewDefinitionId={viewDefinitionId}
      />).instance() as Viewport;

      expect(viewport.selectionHandler.rulesetId).to.eq(rulesetId);
      expect(viewport.selectionHandler.imodel).to.eq(imodelMock.object);
    });

    it("throws when trying to access on component that's not mounted", () => {
      const viewport = shallow(<Viewport
        imodel={imodelMock.object}
        rulesetId={faker.random.word()}
        viewDefinitionId={viewDefinitionId}
        selectionHandler={selectionHandlerMock.object}
      />, { disableLifecycleMethods: true }).instance() as Viewport;
      expect(() => viewport.selectionHandler).to.throw();
    });

    it("disposes when component unmounts", () => {
      const viewport = shallow(<Viewport
        imodel={imodelMock.object}
        rulesetId={faker.random.word()}
        viewDefinitionId={viewDefinitionId}
        selectionHandler={selectionHandlerMock.object}
      />);
      viewport.unmount();
      selectionHandlerMock.verify((x) => x.dispose(), moq.Times.once());
    });

    it("handles missing handler when unmounts", () => {
      const viewport = shallow(<Viewport
        imodel={imodelMock.object}
        rulesetId={faker.random.word()}
        viewDefinitionId={viewDefinitionId}
        selectionHandler={selectionHandlerMock.object}
      />, { disableLifecycleMethods: true });
      viewport.unmount();
    });

    it("updates handler when component's props change", () => {
      const viewport = shallow(<Viewport
        imodel={imodelMock.object}
        rulesetId={faker.random.word()}
        viewDefinitionId={viewDefinitionId}
        selectionHandler={selectionHandlerMock.object}
      />);

      const imodelMock2 = moq.Mock.ofType<IModelConnection>();
      const rulesetId2 = faker.random.word();

      viewport.setProps({
        imodel: imodelMock2.object,
        rulesetId: rulesetId2,
      });

      selectionHandlerMock.verify((x) => x.imodel = imodelMock2.object, moq.Times.once());
      selectionHandlerMock.verify((x) => x.rulesetId = rulesetId2, moq.Times.once());
    });

    it("handles missing handler when updates", () => {
      const viewport = shallow(<Viewport
        imodel={imodelMock.object}
        rulesetId={faker.random.word()}
        viewDefinitionId={viewDefinitionId}
        selectionHandler={selectionHandlerMock.object}
      />, { disableLifecycleMethods: true });
      viewport.instance().componentDidUpdate!(viewport.props(), viewport.state());
    });

  });

});
