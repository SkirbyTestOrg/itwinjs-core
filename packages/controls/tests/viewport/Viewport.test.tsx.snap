// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Viewport renders correctly 1`] = `
<ViewportComponent
  imodel={null}
  viewDefinitionId={"0x18586000000a2e6"}
/>
`;
